### PR TITLE
Image Entities: fix keepAspectRatio, implement naturalDimensions

### PR DESCRIPTION
**Pull request by HifiExperiments**
fix https://github.com/vircadia/vircadia/issues/697

before, keepAspectRatio only worked if the entity dimensions were 1x1.  now, if keepAspectRatio is true, the image will properly fit into its bound no matter what they are.  additionally, image entities now have naturalDimensions.

I couldn't reproduce the part of the bug about keepAspectRatio not working for pngs.

one open question: when a sub image is selected, should it respect the aspect ratio of the original image, or the aspect ratio of the sub image?  it's hard for me to tell which is the correct behavior

Test plan:
- for a variety of images, edit them in create
- confirm that when keepAspectRatio is false, they behave as before, and completely fill their bounds
- confirm that when keepAspectRatio is true, they always show at their original, unstretched aspect ratio, even as you change their dimensions
- make sure png and jpgs behave the same
- press Reset Dimensions.  regardless of if keepAspectRatio is checked, the image should snap to the correct size

### CODE REVIEW

File: libraries/entities-renderer/src/RenderableImageEntityItem.cpp [hunk 1]
```diff
@@ -59,10 +59,22 @@ void ImageEntityRenderer::doRenderUpdateAsynchronousTyped(const TypedEntityPoint
...
+        if (nextTextureLoaded) {
+            float width = _texture->getWidth();
+            float height = _texture->getHeight();
+            glm::vec3 naturalDimensions = glm::vec3(1.0f, 1.0f, 0.01f);
+            if (width < height) {
+                naturalDimensions.x = width / height;
+            } else {
+                naturalDimensions.y = height / width;
+            }
+            entity->setNaturalDimension(naturalDimensions);
```
Comment (HifiExperiments) [inline libraries/entities-renderer/src/RenderableImageEntityItem.cpp:74]: this is kinda hacky.  models handle this a little cleaner because the renderable actually IS the game object, which is itself a hack.  instead of messing with all the image entity code and changing it around, we can sneakily set the dimensions on the game object even though it’s const by using a mutable

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1050/Vircadia-PR1050-e6fe420.tar_5677b5d5.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1050/Vircadia-PR1050-e6fe420.tar_5677b5d5.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1050/Vircadia-PR1050-e6fe420_5677b5d5.sh

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1050/Vircadia-PR1050-e6fe420_5677b5d5.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1050/Vircadia-PR1050-e6fe420_5677b5d5.exe

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1050/Vircadia-PR1050-e6fe420_5677b5d5.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1050/Vircadia-Quest-Alpha-PR1050-e6fe420_5677b5d5.apk

**Comment by Aitolda** [comment]
Tested with jpg, png, bmp, and gif.  All seem to be working as described above.

File: libraries/entities-renderer/src/RenderableImageEntityItem.cpp [hunk 2]
```diff
@@ -59,10 +59,22 @@ void ImageEntityRenderer::doRenderUpdateAsynchronousTyped(const TypedEntityPoint
     _alpha = entity->getAlpha();
     _pulseProperties = entity->getPulseProperties();
 
+    bool nextTextureLoaded = _texture && (_texture->isLoaded() || _texture->isFailed());
     if (!_textureIsLoaded) {
         emit requestRenderUpdate();
```
Comment (ctrlaltdavid) [inline libraries/entities-renderer/src/RenderableImageEntityItem.cpp:64]: Should the `emit requestRenderUpdate()` be done after the `if { ... }`?

**Review by ctrlaltdavid — CHANGES_REQUESTED** [review]

File: libraries/entities-renderer/src/RenderableImageEntityItem.cpp (hunk 2 above)
Comment (HifiExperiments) [reply libraries/entities-renderer/src/RenderableImageEntityItem.cpp:64]: it doesn’t need to be!  the emit just tells the EntityTreeRenderer we need to update again later, we don’t necessarily need to do it last or anything

**Review by ctrlaltdavid — APPROVED** [review]

**Comment by SilverfishVR** [comment]
testing PR1050-e6fe420
generally seems to work as advertised
but I wonder, if I turn on keep aspect on the default image "https://cdn-1.vircadia.com/us-e-1/Bazaar/Assets/Textures/Defaults/Interface/default_image.jpg" it actually looks squished:
Screenshot 2021-03-04 230533

Not sure why that is. No other issues found.

**Comment by SilverfishVR** [comment]
Did some more testing and here is what I have found:

the aspect of the default image is ≈ 0,529:1 (1024x542 pixels) and the dimensions of the image entity when created is ≈ 0.523:1 (0.5385x0.2819 m), close enough, but, the `naturalDimensions` of the new image entity is 0.625:1 (1x0.625 m), even after resetting dimensions.

I thought that resetting dimensions would give it natural dimensions that corrosponds to the aspect ratio of the image but that seems to not be the case, instead they are batched so that all images with an aspect from 0.5:1 to 1:1 get the naturalDimensions of 0.625:1 🤔

So a 1024x512 image is perfect (natural dimensions of 0.5:1) but a 1024x513 image is very squished (natural dimensions of 0.625:1).

Basically `naturalDimensions` is always rounded to 0.5, 0.625, 0.75, 1, 1.25 etc. is that by design?

Here are some of my test textures, jpeg and png seems to behave the same so only png:
[checkerTextures](https://s3.console.aws.amazon.com/s3/buckets/silverfish-freestuff?region=eu-north-1&prefix=TestStuff/checkerTextures/)

**Comment by digisomni** [comment]
This PR is awaiting a fix by @HifiExperiments.

**Comment by HifiExperiments** [comment]
ok I think I've fixed the above issue.  it was a bit involved: when we load a texture, we process it into a GPU-friendly format, which involves messing with its dimensions to make them align better.  during this process, we were losing the original information.  so I've had to pass that around in a number of places, including saving it in the KTX cache.  this will invalidate everyone's KTX cache the first time they load any domain again, but it will be a one time cost.

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1050/Vircadia-PR1050-d9f378e-_35e836d3.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1050/Vircadia-PR1050-d9f378e-.tar_35e836d3.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1050/Vircadia-PR1050-d9f378e-.tar_35e836d3.Z

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1050/Vircadia-PR1050-d9f378e-_35e836d3.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1050/Vircadia-PR1050-d9f378e-_35e836d3.exe

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1050/Vircadia-Quest-Alpha-PR1050-d9f378e_35e836d3.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1050/Vircadia-PR1050-d9f378e_35e836d3.apk

**Comment by SilverfishVR** [comment]
After Clearing the KTX cache it does indeed seem to be working as expected on all my test images 👍 

Note, some textures where indeed corrupted on startup so I did a _"reload content"_ that did not help so I did _"Developer->Render->clear KTX cache"_ and restarted, it worked but I don't know if restarting after "reload content" would have been enough, does that also clear KTX cache? The main difference being that a user would have to enable the developer menu to make it work.
For the release, could the installer clear the cache so it works on first run?

**Comment by HifiExperiments** [comment]
hm I bumped a version number that I *thought* should invalidate the cache automatically...we should be able to make it happen without people having to clear the cache (reload content isn’t enough).  I’ll have to take another look

**Comment by Aitolda** [comment]
Seems like that fixed it, but Silverfish seems to have a better eye for this.

**Comment by daleglass** [comment]
Dev meeting: possible way of testing: Use stable version, copy cache to the PR build's directory, launch PR build

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1050/Vircadia-PR1050-5f4f905-_e5f108b0.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1050/Vircadia-PR1050-5f4f905-.tar_e5f108b0.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1050/Vircadia-PR1050-5f4f905-.tar_e5f108b0.gz

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1050/Vircadia-PR1050-5f4f905-_e5f108b0.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1050/Vircadia-PR1050-5f4f905-_e5f108b0.exe

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1050/Vircadia-Quest-Alpha-PR1050-5f4f905_e5f108b0.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1050/Vircadia-PR1050-5f4f905_e5f108b0.apk

**Review by ctrlaltdavid — CHANGES_REQUESTED** [review]

File: libraries/entities-renderer/src/RenderableImageEntityItem.cpp (hunk 1 above)
Comment (HifiExperiments) [reply libraries/entities-renderer/src/RenderableImageEntityItem.cpp:76]: done!

**Review by ctrlaltdavid — APPROVED** [review]

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1050/Vircadia-PR1050-7bed6cd-_6b82c7fe.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1050/Vircadia-PR1050-7bed6cd-.tar_6b82c7fe.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1050/Vircadia-PR1050-7bed6cd-.tar_6b82c7fe.Z

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1050/Vircadia-PR1050-7bed6cd-_6b82c7fe.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1050/Vircadia-PR1050-7bed6cd-_6b82c7fe.exe

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1050/Vircadia-Quest-Alpha-PR1050-7bed6cd_6b82c7fe.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1050/Vircadia-PR1050-7bed6cd_6b82c7fe.apk

**Comment by SilverfishVR** [comment]
Testing PR-1050-7bed6cd

Tested by copying the "ktx_cache" folder from my release install to the PR install. I did not see any corrupted textures but the default image still behaves weirdly, It has the natural dimensions of `{"x":1,"y":null,"z":0.009999999776482582}` so neither "Keep Aspect Ratio" or "Reset Dimensions" work. As soon as the image is changed it's all good.

If I do the "Reset KTX cache" it works as expected so it really is a one time issue so maybe it is acceptable 🤷‍♂️. Or cheat, by using a different default image in the release :-)
